### ISSUE TITLE
Custom objective function with external variables.

### ISSUE THREAD

**Issue by pavitrakumar78** [fchollet/keras#3029]
Hi,

I am trying to implement a multi layer NN with a custom loss function.

Suppose I have the following function,

```
def custom_loss(y_true,y_pred):  

    loss = T.mean(T.square(y_pred - y_true), axis=-1)  

    return loss  

```
(which is just the default MSE code)
 If I want to add a variable or a function of a few variables that are taken from the outside the NN - is it possible to do so?

for example:

```
def custom_loss(y_true,y_pred):  

    loss = T.mean(T.square(y_pred - y_true), axis=-1) + T.dot(a1,a2)  

    return loss  

```
where a1,a2 are defined outside.

**Comment by soloice** [comment]
Well, you can put these variables into the label y_true.  This might be a hack, but it works.  I don't know if there are more canonical ways.

**Comment by pavitrakumar78** [comment]
@soloice  But, I thought the y_true and y_pred were passed in by the built-in functions of the Model class. Also, the function actually returns only the tensor expression right? not the actual calculated value, since the function is only called once.

**Comment by lemuriandezapada** [comment]
if you use the functional api, you can simply make a new input and directly reference it in your loss.
a1 = Input(whatever_dimensions_1)
a2 = Input(whatever_dimensions_2)
and then directly call them into a loss function

**Comment by nitish11** [comment]
@lemuriandezapada : :+1: 

@pavitrakumar78  :
You can also pass variable using theano shared variables, check the link below : http://deeplearning.net/software/theano/tutorial/examples.html#using-shared-variables

On the other hand, you can also try passing your defined variable tensor symbolic variable to the function.
For Example  :
```

import theano.tensor as tensor

#definng tensor scaler 
x_input = tensor.dscalar('x') # Input

#defining theano function 
f_train = theano.function([x_input], output)

#using the defined variable
x_defined = 25 
f_train(x_defined, out)

```

**Comment by pavitrakumar78** [comment]
@nitish11  and @lemuriandezapada  Thanks for your replies!

**Comment by sharthadavanne** [comment]
so which of the methods worked for you @pavitrakumar78 ?

**Comment by pavitrakumar78** [comment]
@sharthadavanne The method @nitish11 suggested worked well for my purpose.I just passed in whatever variables I needed to the function.

**Comment by nitish11** [comment]
@sharathadavanne : try the following solution.

```
import theano.tensor as tensor

#definng tensor scaler 
x_input = tensor.dscalar('x') # Input

#defining theano function 
f_train = theano.function([x_input], output)

#using the defined variable
x_defined = 25 
f_train(x_defined, out)
```
